### DISCUSSION TITLE
Failure to parse simple Json

### DISCUSSION THREAD

**Discussion by benstadin** [simdjson/simdjson#1993] [General]
I'm a first time simdjson user. I understand the need for padding, yet I do get a weird error on every second or so build without any changes to code (which makes me believe it must have to do something with padding or similar) when I try to get the document as a object root via my getRootAsObject() method below.

Is there something obvious which I'm missing? Or is this indeed a bug?

Simdjson version is 3.1.7 from homebrew with Xcode 14.3 and Apple Clang, GNU++20.

The error I get is this:

`Failed to get JSON document root as object: INCORRECT_TYPE: The JSON element does not have the requested type.`

My code:
```
explicit JsonFile(const std::string &filename) {
        std::ifstream file(filename.c_str());
        if (!file.is_open()) {
            std::string what = "Failed to read JSON file: " + filename;
            throw MyException(what);
        }

        // Load the the file into a buffer
        std::string buffer;
        file.seekg(0, std::ios::end);
        // simdjson braucht padding: https://github.com/simdjson/simdjson/blob/master/doc/basics.md
        size_t size = size_t(file.tellg());
        buffer.resize(size);
        file.seekg(0, std::ios::beg);
        file.read(&buffer[0], file.tellg());
        
        // Store into padded json
        simdjson::padded_string json = buffer;
        simdjson::error_code error = _parser.iterate(json).get(_document);
        if (error != simdjson::SUCCESS) {
            std::string errStr = simdjson::error_message(error) ? std::string(simdjson::error_message(error)) : "";
            std::string what = "Failed to load json document: " + errStr;
            throw MyException(what);
        }
    }

    void getRootAsObject(simdjson::ondemand::object& result) {
        simdjson::error_code error = _document.get_object().get(result);
        if (error != simdjson::SUCCESS) {
            std::string errStr = simdjson::error_message(error) ? std::string(simdjson::error_message(error)) : "";
            std::string what = "Failed to get JSON document root as object: " + errStr;
            throw MyException(what);
        }
    }
```

**Comment by lemire** [comment]
You should be able to greatly simplify your code with a line such as...

```C++
json = padded_string::load(filename); 
```

> Is there something obvious which I'm missing? 


1. The file is read but never closed. You can fix that by using the line above...
2. After the `JsonFile` constructor, the `simdjson::padded_string json` instance which contains the input string is destroyed, and the memory is released. It is possible, just out of chance, that access to it might remain safe for a while, even after the destructor has been called (so your code might work some of the time), but it is certainly not *safe* and it will definitively crash some of the time. I refer you to our documentation which states that the input string must remain available during the parsing:
<img width="1062" alt="Capture d’écran, le 2023-04-23 à 11 44 21" src="https://user-images.githubusercontent.com/391987/233849793-e6134256-2a6e-4afd-9801-2ae27f2c848a.png">


Please note that simdjson has been used in important production systems for many years. I am sure we still have bugs, and some are discovered from time to time, but the overwhelming majority of the library is bug-free at this point. We also benefit from extensive testing.
